On Parallel Binary Search

Kostero opened the pull request.

mhayter commented:
Hello @Kostero and welcome! Thanks for joining the project and thanks for the contribution! 

Here's some quick initial feedback:

Consider putting text in online grammar/spell checker. I noticed than 'Specifically' was misspelled. 

Also, I'd personally prefer more descriptive variable names rather than `A`, `X`, `M`, etc. especially consider `M` is already used in the article for midpoint.

Also, consider compiling the given code.

What is `P` that is returned?  

Also, I think we use snake case for functions and it may make sense to have `left` and `right` be a `struct` as they are parallel arrays.

Kostero commented:
Thanks for the feedback. 

I have fixed some issues. Comments for the remaining ones below.

> Also, I'd personally prefer more descriptive variable names rather than A, X, M, etc. especially consider M is already used in the article for midpoint.

A and X are mostly there to keep things simple and to not repeat long variables names (especially in the table explaining what we actually do). I would prefer to keep it that way.

> What is P that is returned?

I have a problem of changing the variables over and over, after doing all the checks (including compilation). It should work now. I will try to add tests in the follow-up, just wanted to get the first review asap.

> It may make sense to have left and right be a struct as they are parallel arrays.

I kinda disagree here, as they directly refer to `int l = -1, r = n;` in the prior binary search code (I changed the variable names now to be more consistent), so I would keep them as separate parallel arrays (as they are separate parallel variables in that code).

adamant-pwn commented on src/num_methods/binary_search.md:
```diff
@@ -138,6 +138,63 @@ Another noteworthy way to do binary search is, instead of maintaining an active
 
+## Parallel Binary Search
+
+<small>Note that this section follows the description in [Sports programming in practice](https://kostka.dev/sp/).</small>
```
Is the intention here to provide a reference for further reading, or an attribution? I think it is more common to integrate references in the text (see e.g. how [this](https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html) is linked above in the article) or put them in some kind of further reading section at the end.

Also, to make sure, you understand that by putting the text from the book here you also make it licensed under CC BY-SA 4.0?

adamant-pwn commented on src/num_methods/binary_search.md:
```diff
@@ -138,6 +138,63 @@ Another noteworthy way to do binary search is, instead of maintaining an active
+| **step 4** | answer in \([3,4)\)   | answer in \([5,6)\)   | answer in \([1,2)\)  | answer in \([2,3)\)  |
+|        | \( index = 3 \)       | \( index = 5 \)       | \( index = 1 \)      | \( index = 2 \)      |
+
+We generally process this table by columns (queries), but notice that in each row we often repeat access to certain values of the array. To limit access to these values, we can process the table by rows (steps). This does not make huge difference in our small example problem (as we can access all elements in $\mathcal{O}(1)$), but in more complex problems, where computing these values is more com ...
```
> Moreover, note that we can arbitrarily choose the order in which we answer questions in a single row.

Don't we actually really care about doing it in increasing order of $m$ in certain scanline-like applications?

adamant-pwn reviewed (changes requested):
Thanks for the pull request! I have a few comments that I think should be addressed before this is merged.

mhayter commented:
Any update here?

mhayter commented:
Is this dead?

Kostero commented on src/num_methods/binary_search.md:
```diff
@@ -138,6 +138,63 @@ Another noteworthy way to do binary search is, instead of maintaining an active
+| **step 4** | answer in \([3,4)\)   | answer in \([5,6)\)   | answer in \([1,2)\)  | answer in \([2,3)\)  |
+|        | \( index = 3 \)       | \( index = 5 \)       | \( index = 1 \)      | \( index = 2 \)      |
+
+We generally process this table by columns (queries), but notice that in each row we often repeat access to certain values of the array. To limit access to these values, we can process the table by rows (steps). This does not make huge difference in our small example problem (as we can access all elements in $\mathcal{O}(1)$), but in more complex problems, where computing these values is more com ...
```
PTAL now
